Autonomize HTML button additions using JS

ccullip opened the pull request:
Function that adds buttons to the HTML using javascript and array of dictionary keys

ccullip committed:
Autonomize button additions

ccullip committed:
Merge pull request #1 from ccullip/block_edits

Autonomize button additions

ccullip committed:
Remove unneccessary <div>

justinnhli commented on brick-editor.js:
```diff
@@ -1,6 +1,18 @@
+    'background-color:coral',
+    'background-color:orange']
+
+var keysArray = ['IF', 'IF-ELSE', 'FOR', 'WHILE', 'VARIABLE', 'FUNCTION'];
```
These would require manual updating, which is not ideal. See next comment.

justinnhli replied:
Both of these arrays would require manual updating. See comments below.

justinnhli commented on brick-editor.js:
```diff
@@ -63,19 +75,43 @@ function addBlock(word) {
+        'FOR': "for (var i = 0; i < value; i++){\n" + tabs + "\t // do something \n" + tabs + "}",
+        'WHILE': "while (i < 10) {\n" + tabs + "\t" + "// do something \n" + tabs + "}",
+        'FUNCTION': "function name(parameters) {\n" + tabs + "\t // do something \n" + tabs + "\t return value;\n" + tabs + "}",
     };
```
Instead, the better way to do this is to derive the keys from this dictionary. This might mean changing this from a dictionary to a nested list (so the key-order is preserved).

For the colors, you might pre-define some number of colors and pick from those. Alternately, you might hard-code the colors into the dictionary as well. I can see the latter approach being useful if the buttons will be in groups (ie. control flow vs. functions).

This means your "dictionary" might end up being a nested list like:

```
var blockDict = [
    ['var', 'var variableName = value;', '#ff3399'],
    // ...
];
```

justinnhli replied:
Or actually, even better would be to make this a list of JSON objects:

```javascript
var blockDict = [
    {
        "block-name":"var",
        "code":"var variableName = value;",
        "button-color":"#ff3399"
    },
    // ...
];
```

ccullip replied:
is it better practice to use hexadecimal instead of the english color names?

justinnhli replied:
Yes, I would say so. You can include the English name as a comment if you want, but I'm not convinced it's that helpful anyway.

justinnhli commented on brick-editor.js:
```diff
@@ -63,19 +75,43 @@ function addBlock(word) {
+        block.appendChild(text);
+
+        // adds the new button inside the buttonContainer class at end
+        var existingElement = document.getElementById("buttonContainer");
```
Rename this to `buttonContainer`, to match the HTML id.

justinnhli commented on brick-editor.js:
```diff
@@ -63,19 +75,43 @@ function addBlock(word) {
+        block.setAttribute("style", colorArray[i]);
+        block.setAttribute("onclick", clickEvent);
+        var text = document.createTextNode(keysArray[i]);
+        block.appendChild(text);
```
No need for the extra `text` variable here, just do it inline:

`block.appendChild(document.createTextNode(keysArray[i]))`

justinnhli commented on brick-editor.js:
```diff
@@ -63,19 +75,43 @@ function addBlock(word) {
+
+        // adds a break element to make a column of blocks
+        var br = document.createElement("br");
+        existingElement.appendChild(br);
```
Again, no need for a temporary variable here.

justinnhli commented on brick-editor.js:
```diff
@@ -63,19 +75,43 @@ function addBlock(word) {
 
+// adds all the blocks to the button container
+function addBlocksHTML() {
+    for (i in keysArray) {
```
This loop will have to change to accommodate the dictionary changes above. Using `for...in` in Javascript is also a bad idea. See https://stackoverflow.com/a/9329476

ccullip replied:
Can I use a normal for loop in place of `for... in` then?
Like `for (var i = 0; i < array.length; i++) {
...
}`

justinnhli replied:
Yes, I agree with this approach.

justinnhli commented on brick-editor.js:
```diff
@@ -63,19 +75,43 @@ function addBlock(word) {
+// adds all the blocks to the button container
+function addBlocksHTML() {
+    for (i in keysArray) {
+        var clickEvent = 'addBlock("' + keysArray[i] + '")';
```
No need for the `clickEvent` variable here. Just do it inline.

justinnhli reviewed (changes requested):
Where possible, derive everything from the dictionary, including keys and color. Some JavaScript temporary variables can also be eliminated.

justinnhli commented on brick-editor.js:
```diff
@@ -63,19 +75,43 @@ function addBlock(word) {
+        block.setAttribute("type", "button");
+        block.setAttribute("class", "addButton");
+        block.setAttribute("style", colorArray[i]);
+        block.setAttribute("onclick", clickEvent);
```
Instead of using `setAttribute`, use `onclick`: https://www.w3schools.com/jsref/tryit.asp?filename=tryjsref_onclick_dom

ccullip committed:
Change BlockDict to JSON objects

Has error with tab indentation due to local/global variable issues. Attempt will be made using recast to reformat code with correct indentation

justinnhli commented:
PS. Please leave a comment on the pull request itself when the code is ready for review again.

ccullip commented:
For fixing the undefined issue, is there a way to "reload" the dictionary keys after finding the number of tabs necessary in getIndent() ? Or would I have to redesign getIndent() to mess directly with the tab indentation by splitting at \n and then rejoining with correct number of tabs?

justinnhli commented:
...neither? Here's what I did (feel free to copy, test, and commit):

```javascript
function indentCode(code, tabs) {
    var lines = code.split('\n');
    for (var i = 1; i < lines.length; i++) {
        lines[i] = tabs + lines[i];
    }
    return lines.join('\n');
}

// adds a block based on word
function addBlock(i, word) {
    var tabs = getIndent(position);
    var buffer = editor.getValue();
    var firstPart = getBeforeCursor(buffer, position);
    var lastPart = getAfterCursor(buffer, position);

    var indentedCode = indentCode(blockDict[i].code, tabs)
    var block = [firstPart, indentedCode, lastPart].join("");
    editor.setValue(block);
    editor.setPosition(position);
}
```

justinnhli commented:
And you can take out the tabs entirely from the templates:

```javascript
// initialize dictionary
var blockDict = [
    {
        'blockName': 'IF',
        'code': 'if (i == true) {\n\t// do something\n}',
        'buttonColor': '#ff3399', // fuschia
    },
    {
        'blockName': 'IF-ELSE',
        'code': 'if (i == true) {\n\t// do something\n} else {\n\t// do something\n}',
        'buttonColor': '#b8860b', // darkgoldenrod
    },
    {
        'blockName': 'FOR',
        'code': 'for (var i = 0; i < value; i++){\n\t // do something\n}',
        'buttonColor': '#00bfff', // deepskyblue
    },
    {
        'blockName': 'WHILE',
        'code': 'while (i < 10) {\n\t// do something\n}',
        'buttonColor': '#32cd32' // lime green
    },
    {
        'blockName': 'VARIABLE',
        'code': 'var variableName = value;',
        'buttonColor': '#9932cc' // darkorchid
    },
    {
        'blockName': 'FUNCTION',
        'code': 'function name(parameters) {\n\t // do something\n\t return value;\n}',
        'buttonColor': '#ff7f50' // coral
    },
];
```

ccullip commented:
Okay. I found that this works: 
`function getIndent(position, code) {
    tabs = "";
    for (var i = 0; i < position.column - 2; i = i + 4) {
        tabs += "\t";
    }
    var codeArray = code.split("\n");
    for (var i = 1; i < codeArray.length; i++) {
        codeArray[i] = tabs.concat(codeArray[i]);
    }
    code = codeArray.join("\n");
    return code;
}

// adds a block based on word
function addBlock(i, word) {
    var buffer = editor.getValue();
    var firstPart = getBeforeCursor(buffer, position);
    var lastPart = getAfterCursor(buffer, position);

    var block = [firstPart, getIndent(position, blockDict[i]["code"]), lastPart].join("");
    editor.setValue(block);
    editor.setPosition(position);`

It's a bit longer than yours, though.

ccullip commented:
Okay. I found that this works, although it's a bit longer. I also deleted tabs from the dictionary.
``` javascript
function getIndent(position, code) {
    tabs = "";
    for (var i = 0; i < position.column - 2; i = i + 4) {
        tabs += "\t";
    }
    var codeArray = code.split("\n");
    for (var i = 1; i < codeArray.length; i++) {
        codeArray[i] = tabs.concat(codeArray[i]);
    }
    code = codeArray.join("\n");
    return code;
}

// adds a block based on word
function addBlock(i, word) {
    var buffer = editor.getValue();
    var firstPart = getBeforeCursor(buffer, position);
    var lastPart = getAfterCursor(buffer, position);

    var block = [firstPart, getIndent(position, blockDict[i]["code"]), lastPart].join("");
    editor.setValue(block);
    editor.setPosition(position);
```

justinnhli commented:
Same idea, I think - as long as you take the tabs to indent the template, you should be fine. I like having the `indentCode()` as a separate function, but it's up to you which one to commit. Push to your branch and I'll take a final look?

PS: you can write three backticks then a language (eg. <code>```javascript</code>) to start a multi-line code block with language-specific highlighting. End the block with another three backticks.

ccullip committed:
Fix "undefined" tabs issue

ccullip committed:
Fix "undefined" tabs issue

justinnhli commented on brick-editor.js:
```diff
@@ -1,6 +1,41 @@
 var editor;
 var position;
-var blockDict = {};
+var tabs = "";
```
Delete this line; tabs should be a local variable from now on.

justinnhli commented on brick-editor.js:
```diff
@@ -45,7 +80,7 @@ function start_brick_editor() {
-var tabs;
+
 function getIndent(position) {
     tabs = "";
```
This line should be `var tabs = "";`

justinnhli commented on brick-editor.js:
```diff
@@ -54,28 +89,48 @@ function getIndent(position) {
+        codeArray[i] = tabs.concat(codeArray[i]);
+    }
+    code = codeArray.join("\n");
+    return code;
```
Just directly return the string: `return codeArray.join("\n");`

ccullip committed:
Merge remote-tracking branch 'origin/master'

justinnhli reviewed (commented):
Three small changes, and then we'll be done. Thanks for entertaining my fussiness.

ccullip commented:
So when there are small changes like deleting a line or two, do I do a whole commit or is the "amend" commit thing better?

ccullip committed:
Make tabs a local variable and small code clean-up

justinnhli commented:
To answer your question: it's entirely up to you. I won't enforce it for this class, but you can go back and clean up your commits to make things neater as well, by doing a rebase and reordering your commits. Feel free to come by office hours if you want to learn more.
